File: software/groveapi/internal/logic/availability.go
package logic

import (
	"context"
	"encoding/json"
	"sort"
	"time"

	"groveapi/internal/store"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Slot struct {
	StartsAt  time.Time `json:"starts_at"`
	EndsAt    time.Time `json:"ends_at"`
	Available bool      `json:"available"`
}

type AvailabilityResult struct {
	ResourceID string `json:"resource_id"`
	From       string `json:"from"`
	To         string `json:"to"`
	Slots      []Slot `json:"slots"`
}

type DayHours struct {
	Open  string `json:"open"`
	Close string `json:"close"`
}

func GetAvailability(ctx context.Context, db *gorm.DB, resourceID uuid.UUID, from, to time.Time) (AvailabilityResult, error) {
	var resource store.Resource
	if err := db.WithContext(ctx).Where("id = ?", resourceID).First(&resource).Error; err != nil {
		return AvailabilityResult{}, ErrNotFound
	}

	// Parse open hours
	openHours := map[string]DayHours{}
	if len(resource.OpenHours) > 0 && string(resource.OpenHours) != "{}" {
		_ = json.Unmarshal(resource.OpenHours, &openHours)
	}

	// Get existing reservations in range
	var reservations []store.Reservation
	db.WithContext(ctx).
		Where("resource_id = ? AND ends_at > ? AND starts_at < ? AND status IN ?",
			resourceID, from, to, []string{"HELD", "CONFIRMED"}).
		Find(&reservations)

	slotDuration := time.Duration(resource.SlotMinutes) * time.Minute
	bufferDuration := time.Duration(resource.BufferMinutes) * time.Minute

	var slots []Slot

	// Generate slots day by day
	for day := from; day.Before(to); day = day.AddDate(0, 0, 1) {
		dayName := dayNameLower(day.Weekday())
		hours, ok := openHours[dayName]
		if !ok {
			// Default: 06:00-22:00 if no open hours defined
			hours = DayHours{Open: "06:00", Close: "22:00"}
		}

		openTime, err := time.Parse("15:04", hours.Open)
		if err != nil {
			continue
		}
		closeTime, err := time.Parse("15:04", hours.Close)
		if err != nil {
			continue
		}

		dayStart := time.Date(day.Year(), day.Month(), day.Day(),
			openTime.Hour(), openTime.Minute(), 0, 0, day.Location())
		dayEnd := time.Date(day.Year(), day.Month(), day.Day(),
			closeTime.Hour(), closeTime.Minute(), 0, 0, day.Location())

		for slotStart := dayStart; slotStart.Add(slotDuration).Before(dayEnd) || slotStart.Add(slotDuration).Equal(dayEnd); slotStart = slotStart.Add(slotDuration + bufferDuration) {
			slotEnd := slotStart.Add(slotDuration)
			available := true
			for _, r := range reservations {
				if r.StartsAt.Before(slotEnd) && r.EndsAt.After(slotStart) {
					available = false
					break
				}
			}
			slots = append(slots, Slot{
				StartsAt:  slotStart,
				EndsAt:    slotEnd,
				Available: available,
			})
		}
	}

	sort.Slice(slots, func(i, j int) bool {
		return slots[i].StartsAt.Before(slots[j].StartsAt)
	})

	return AvailabilityResult{
		ResourceID: resourceID.String(),
		From:       from.Format(time.RFC3339),
		To:         to.Format(time.RFC3339),
		Slots:      slots,
	}, nil
}

func dayNameLower(d time.Weekday) string {
	names := [7]string{"sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"}
	return names[d]
}

File: software/clients/web/src/lib/availability.ts
import { supabase } from "./supabase"
import type { Resource, Reservation } from "./types"

export type Slot = {
  starts_at: string
  ends_at: string
  available: boolean
  booked_by_me: boolean
}

type DayHours = {
  open: string
  close: string
}

const DAY_NAMES = ["sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"]

function parseHHMM(hhmm: string): { hour: number; minute: number } | null {
  const [h, m] = hhmm.split(":").map(Number)
  if (isNaN(h) || isNaN(m)) return null
  return { hour: h, minute: m }
}

export async function getAvailability(
  resourceId: string,
  from: Date,
  to: Date,
): Promise<{ resource_id: string; from: string; to: string; slots: Slot[] }> {
  // Fetch resource, overlapping reservations, and current user in parallel
  const [resourceResult, reservationsResult, userResult] = await Promise.all([
    supabase.from("resources").select("*").eq("id", resourceId).single(),
    supabase
      .from("reservations")
      .select("starts_at, ends_at, user_id")
      .eq("resource_id", resourceId)
      .in("status", ["HELD", "CONFIRMED"])
      .lt("starts_at", to.toISOString())
      .gt("ends_at", from.toISOString()),
    supabase.auth.getUser(),
  ])

  if (resourceResult.error) throw resourceResult.error
  const resource = resourceResult.data as Resource
  const reservations = (reservationsResult.data ?? []) as Pick<Reservation, "starts_at" | "ends_at" | "user_id">[]
  const currentUserId = userResult.data.user?.id

  // Parse open_hours
  const openHours: Record<string, DayHours> = typeof resource.open_hours === "object" && resource.open_hours !== null
    ? (resource.open_hours as unknown as Record<string, DayHours>)
    : {}

  const slotMs = resource.slot_minutes * 60_000
  const bufferMs = resource.buffer_minutes * 60_000

  const slots: Slot[] = []

  // Generate slots day by day
  const day = new Date(from)
  while (day < to) {
    const dayName = DAY_NAMES[day.getDay()]
    const hours = openHours[dayName] ?? { open: "06:00", close: "22:00" }

    const openTime = parseHHMM(hours.open)
    const closeTime = parseHHMM(hours.close)
    if (!openTime || !closeTime) {
      day.setDate(day.getDate() + 1)
      continue
    }

    const dayStart = new Date(day)
    dayStart.setHours(openTime.hour, openTime.minute, 0, 0)

    const dayEnd = new Date(day)
    dayEnd.setHours(closeTime.hour, closeTime.minute, 0, 0)

    let slotStart = dayStart.getTime()
    while (slotStart + slotMs <= dayEnd.getTime()) {
      const slotEnd = slotStart + slotMs

      let available = true
      let booked_by_me = false
      for (const r of reservations) {
        const rStart = new Date(r.starts_at).getTime()
        const rEnd = new Date(r.ends_at).getTime()
        if (rStart < slotEnd && rEnd > slotStart) {
          available = false
          if (r.user_id === currentUserId) booked_by_me = true
          break
        }
      }

      slots.push({
        starts_at: new Date(slotStart).toISOString(),
        ends_at: new Date(slotEnd).toISOString(),
        available,
        booked_by_me,
      })

      slotStart += slotMs + bufferMs
    }

    day.setDate(day.getDate() + 1)
  }

  slots.sort((a, b) => a.starts_at.localeCompare(b.starts_at))

  return {
    resource_id: resourceId,
    from: from.toISOString(),
    to: to.toISOString(),
    slots,
  }
}
